### ISSUE TITLE
Improve Serialization Performance

### ISSUE THREAD

**Issue by prestonvanloon** [prysmaticlabs/geth-sharding#139]
I ran some benchmark tests against a few simple round trip serializations for collations and found tha the performance was not that great. I wonder if the bottleneck is within our serialization code, RLP, or something else. 

Any improvements in this area should reference this issue. 
This can be closed once we come to a consensus that the serialization is the absolute most efficent that it could ever be in terms of compute operations and memory allocations.

```
goos: linux
goarch: amd64
pkg: github.com/ethereum/go-ethereum/sharding
BenchmarkSerialization10-4                 20000             84624 ns/op           97347 B/op        565 allocs/op
BenchmarkSerialization100-4                 2000            820836 ns/op          994422 B/op       5528 allocs/op
BenchmarkSerialization1000-4                 200           8076776 ns/op        10630373 B/op      55047 allocs/op
BenchmarkSerialization10000-4                  5         242064047 ns/op        244503252 B/op    350102 allocs/op
PASS
ok      github.com/ethereum/go-ethereum/sharding        9.118s
```

Here's a comparison against other protocols:

Protocol Buffers Marshal: *819 ns/op*
Protocol Buffers Unmarshal: *1163 ns/op*
JSON Marshal: *3316 ns/op*
JSON Unmarshal: *7196 ns/op*
XML Marshal: *9248 ns/op*
XML Unmarshal: *30485 ns/op*
[source](https://medium.com/@shijuvar/benchmarking-protocol-buffers-json-and-xml-in-go-57fa89b8525)
Note: Our benchmarks are roundtrip serialize/deserialize. 

See #138 for original context.

**Comment by nisdas** [comment]
Assigned to [Yutaro](https://github.com/rawfalafel)

**Comment by rauljordan** [comment]
Closed by #138
